File: packages/dart/lib/Seguranca/Ataques/AtaqueCorrelacaoMesmoPlaintext.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/CriptografiaAlvo.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/SkipAtaqueException.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// O AtaqueColisaoIV já garante que o MESMO plaintext nunca reusa o IV.
/// Esse ataque vai além: verifica se os CIPHERTEXTS resultantes, embora
/// venham do mesmo texto, se comportam como se fossem de textos diferentes
/// (distância de Hamming média ~50%, sem nenhum padrão fixo entre eles).
///
/// Se o IV está fazendo seu trabalho, cifrar "SEGREDO" 500 vezes deveria
/// parecer, aos olhos de quem só vê o ciphertext, tão aleatório quanto
/// cifrar 500 textos diferentes.
class AtaqueCorrelacaoMesmoPlaintext extends Ataque {
  final int _amostras;
  final String _textoFixo;

  AtaqueCorrelacaoMesmoPlaintext({
    int amostras = 300,
    String textoFixo = 'MENSAGEM_SEMPRE_IGUAL_PARA_TESTAR',
  }) : _amostras = amostras,
       _textoFixo = textoFixo;

  @override
  String nome() =>
      'Correlação entre ciphertexts do mesmo plaintext';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    if (alvo is! CriptografiaAlvo) {
      throw SkipAtaqueException('Precisa de base64url_decode do alvo.');
    }

    final ciphertexts = <Uint8List>[];
    for (var i = 0; i < _amostras; i++) {
      final token = alvo.encrypt(_textoFixo);
      final decodificado = alvo.base64urlDecode(
        token.substring(alvo.prefixo().length),
      );
      ciphertexts.add(alvo.decompor(decodificado)['ciphertext']!);
    }

    // Compara pares aleatórios de ciphertexts (não todos contra todos,
    // pra manter o custo baixo) e mede a distância de Hamming média.
    final comparacoes =
        _min(500, (_amostras * (_amostras - 1)) ~/ 2);
    final distancias = <double>[];
    for (var c = 0; c < comparacoes; c++) {
      final i = randomInt(0, _amostras - 1);
      final j = randomInt(0, _amostras - 1);
      if (i == j) {
        continue;
      }
      distancias.add(_distanciaHammingRelativa(ciphertexts[i], ciphertexts[j]));
    }

    final media = distancias.reduce((a, b) => a + b) / distancias.length;

    // Também confere que nenhum PAR de ciphertexts é idêntico (o que
    // indicaria reuso de IV+key, capturado de outro ângulo).
    final unicos = ciphertexts.map(bytesToHex).toSet();
    final duplicatas = ciphertexts.length - unicos.length;

    final vulneravel = media < 0.4 || media > 0.6 || duplicatas > 0;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.alta : Severidade.info,
      'distância de Hamming média entre ciphertexts do mesmo texto: '
          '${(media * 100).toStringAsFixed(1)}% (esperado ~50%), '
          '$duplicatas duplicata(s) exata(s) em $_amostras amostras',
      {'media': media, 'duplicatas': duplicatas},
    );
  }

  double _distanciaHammingRelativa(Uint8List a, Uint8List b) {
    final len = _min(a.length, b.length);
    if (len == 0) {
      return 0.5;
    }
    var diff = 0;
    for (var i = 0; i < len; i++) {
      diff += contarBits1(a[i] ^ b[i]);
    }
    return diff / (len * 8);
  }

  int _min(int a, int b) => a < b ? a : b;
}

File: packages/rust/src/seguranca/ataques/ataque_correlacao_mesmo_plaintext.rs
//! O AtaqueColisaoIV já garante que o MESMO plaintext nunca reusa o IV.
//! Esse ataque vai além: verifica se os CIPHERTEXTS resultantes, embora
//! venham do mesmo texto, se comportam como se fossem de textos diferentes
//! (distância de Hamming média ~50%, sem nenhum padrão fixo entre eles).
//!
//! Se o IV está fazendo seu trabalho, cifrar "SEGREDO" 500 vezes deveria
//! parecer, aos olhos de quem só vê o ciphertext, tão aleatório quanto
//! cifrar 500 textos diferentes.

use std::collections::HashSet;

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::criptografia_alvo::CriptografiaAlvo;
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::{contar_bits1, hex_encode, random_int};

pub struct AtaqueCorrelacaoMesmoPlaintext {
    pub amostras: usize,
    pub texto_fixo: String,
}

impl AtaqueCorrelacaoMesmoPlaintext {
    pub fn new() -> Self {
        AtaqueCorrelacaoMesmoPlaintext {
            amostras: 300,
            texto_fixo: "MENSAGEM_SEMPRE_IGUAL_PARA_TESTAR".to_string(),
        }
    }
}

impl Default for AtaqueCorrelacaoMesmoPlaintext {
    fn default() -> Self {
        Self::new()
    }
}

impl AtaqueCorrelacaoMesmoPlaintext {
    fn distancia_hamming_relativa(a: &[u8], b: &[u8]) -> f64 {
        let len = a.len().min(b.len());
        if len == 0 {
            return 0.5;
        }
        let mut diff = 0u32;
        for i in 0..len {
            diff += contar_bits1(a[i] ^ b[i]);
        }
        diff as f64 / (len * 8) as f64
    }
}

impl Ataque for AtaqueCorrelacaoMesmoPlaintext {
    fn nome(&self) -> String {
        "Correlação entre ciphertexts do mesmo plaintext".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        if alvo.as_any().downcast_ref::<CriptografiaAlvo>().is_none() {
            return Err(ErroAtaque::Skip(
                "Precisa de base64url_decode do alvo.".to_string(),
            ));
        }

        let mut ciphertexts: Vec<Vec<u8>> = Vec::with_capacity(self.amostras);
        for _ in 0..self.amostras {
            let token = alvo
                .encrypt(self.texto_fixo.as_bytes())
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            let decodificado = alvo
                .base64url_decode(&token[alvo.prefixo().len()..])
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            ciphertexts.push(alvo.decompor(&decodificado).ciphertext);
        }

        // Compara pares aleatórios de ciphertexts (não todos contra todos,
        // pra manter o custo baixo) e mede a distância de Hamming média.
        let comparacoes = ((self.amostras * (self.amostras - 1)) / 2).min(500);
        let mut distancias: Vec<f64> = Vec::new();
        for _ in 0..comparacoes {
            let i = random_int(0, self.amostras as i64 - 1) as usize;
            let j = random_int(0, self.amostras as i64 - 1) as usize;
            if i == j {
                continue;
            }
            distancias.push(Self::distancia_hamming_relativa(
                &ciphertexts[i],
                &ciphertexts[j],
            ));
        }

        let media = distancias.iter().sum::<f64>() / distancias.len() as f64;

        // Também confere que nenhum PAR de ciphertexts é idêntico (o que
        // indicaria reuso de IV+key, capturado de outro ângulo).
        let unicos: HashSet<String> = ciphertexts.iter().map(|b| hex_encode(b)).collect();
        let duplicatas = ciphertexts.len() - unicos.len();

        let vulneravel = media < 0.4 || media > 0.6 || duplicatas > 0;

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Alta
            } else {
                Severidade::Info
            },
            &format!(
                "distância de Hamming média entre ciphertexts do mesmo texto: {:.1}% (esperado ~50%), {} duplicata(s) exata(s) em {} amostras",
                media * 100.0, duplicatas, self.amostras
            ),
        ))
    }
}
